staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go

```diff
@@ -512,6 +525,36 @@ func PollImmediateUntil(interval time.Duration, condition ConditionFunc, stopCh
 	}
 }
 
+// PollUntilWithContext tries a condition func until it returns true,
+// an error or the specified context is cancelled or expired.
+//
+// PollUntilWithContext always waits interval before the first run of 'condition'.
+// 'condition' will always be invoked at least once.
+func PollUntilWithContext(ctx context.Context, interval time.Duration, condition ConditionWithContextFunc) error {
+	return WaitForWithContext(ctx, poller(interval, 0), condition)
+}
+
+// PollImmediateUntilWithContext tries a condition func until it returns true,
+// an error or the specified context is cancelled or expired.
+//
+// PollImmediateUntilWithContext runs the 'condition' before waiting for the interval.
+// 'condition' will always be invoked at least once.
+func PollImmediateUntilWithContext(ctx context.Context, interval time.Duration, condition ConditionWithContextFunc) error {
+	done, err := condition(ctx)
```

Comment:
I expected to see an adapter for `ConditionFunc` that allows it to match this `ConditionWithContextFunc`. Then I expected to see `PollUntil` call `PollUntilWithContext`, and congruent peers throughout.  That way we don't have duplicated functions like this.

Reply:
A couple of notes:
- today, for functions without `context` we return `ErrWaitTimeout`. for context enabled function it seems more appropriate to return `ctx.Err()`. If we want to maintain backward compatibility we will need to do:
```
	case <-ctx.Done():
		// returning ctx.Err() will break backward compatibility.
		return ErrWaitTimeout
```

- we want to avoid creating additional goroutine as a result of deriving a context from a `<-chan struct{}`. 

I now have a private `poll` function that is used by all public facing `Poll*` functions in this package. Let me know if this looks good.